src/shm/mod.rs

```diff
@@ -1,80 +1,155 @@
-//! Various small utilities helping you to write clients
-
-use std::{cell::RefCell, rc::Rc};
+pub mod pool;
 
 use wayland_client::{
-    protocol::{wl_registry, wl_shm},
-    Attached, DispatchData,
+    protocol::{wl_shm, wl_shm_pool},
+    ConnectionHandle, DataInit, DelegateDispatch, DelegateDispatchBase, Dispatch, QueueHandle,
+    WEnum,
 };
 
-mod mempool;
+use crate::registry::{RegistryHandle, RegistryHandler};
+
+use self::pool::{raw::RawPool, simple::SimplePool, CreatePoolError};
 
-pub use self::mempool::{AutoMemPool, DoubleMemPool, MemPool};
-pub use wl_shm::Format;
+pub trait ShmHandler {
+    /// Indicates the compositor supports an SHM pool with the specified format.
+    fn supported_format(&mut self, format: wl_shm::Format);
+}
```

Comment:
I'm thinking the `ShmState` should instead store the supported formats in an internal `Vec`, and make it available to outside via a method, as was done previously. That's what the user is going to need anyway, no need to use this trait dance here, that is not an event the user is expected to need to answer to.